### ISSUE TITLE
Apparent subtle unsoundness

### ISSUE THREAD

**Issue by btj** [Z3Prover/z3#2244]
I am seeing Z3 behaviour that seems unsound: a `Z3_solver_check` call returns `L_FALSE` where it seemingly should not. I say "seemingly" because the run where this happens is fairly large, with many Z3 API calls, and while the `L_FALSE` result is definitely unexpected, it may be due to bad input to Z3 (i.e. a bug in my tool) rather than a bug in Z3. I am having trouble producing a smaller repro: the phenomenon seems to be very brittle. For example, removing earlier "branches" (stuff between a push and matching pop) seems to make it go away. (I say "seems" because I tried this at the level of the input to my tool (VeriFast) rather than at the level of the Z3 API calls, so there may be something else going on as well.)

The replay log is [z3log.zip](https://github.com/Z3Prover/z3/files/3100964/z3log.zip). It ends with a type error, which is expected; the call that produces the unexpected result is the `Z3_solver_check` call preceding it. It works with 4.8.4; I did not try with HEAD.

I wonder if there is an easy way to tell whether this is in fact a Z3 unsoundness, or whether the problem is in my tool?

**Comment by NikolajBjorner** [comment]
The log says in the first line:

V "4.5.0.0 Dec 14 2017"

Thus, it isn't 4.8.4 but a much older version of Z3. 
Could you create a log using the current head of Z3? It would make it much easier to identify whether this was already a known issue etc.

**Comment by btj** [comment]
Oops! I thought I was using Z3 4.8.4, but I was using the Z3 4.5.0 DLL after all. I now tried to reproduce the problem on Z3 4.8.4, and I failed. Of course, it's possible that there is still an unsoundness in Z3 but that it is now masked by some other phenomenon. But let's be optimistic and assume that either it was not an unsoundness in the first place, or that it was but it has been fixed. So I'm closing this issue. Thanks for having looked at it!
